Add dependency on `pathlib-abc`

barneygale opened the pull request:
Make `zipp.Path` subclass `pathlib_abc.ReadablePath`. This allows us to remove implementations of `read_text()`, `read_bytes()`, `glob()`, `joinpath()` and `__truediv__()`, and to simplify implementations of a couple of few more methods.

Maintain a tree of `PathInfo` objects representing the hierarchy of zip file members. We traverse the tree whenever we need to resolve a path to a `ZipInfo` object. This effectively hides the `.zip`-specific quirk that directories are recorded with a trailing slash in their filenames.

Adjust `__str__()` so that it doesn't raise for unnamed zip files. Instead it returns a string beginning with `:fileobj:`. The prefix is made available as a new `anchor` attribute.

Add the following methods/attributes, which are required by `ReadablePath`:

- `info`
- `parser`
- `__open_rb__()`
- `readlink()`
- `with_segments()`

Disable the following `ReadablePath` methods/attributes that we don't (yet) test in zipp:

- `anchor`
- `parts`
- `parents`
- `__rtruediv__()`
- `with_name()`
- `with_stem()`
- `with_suffix()`
- `full_match()`
- `walk()`
- `copy()`
- `copy_into()`

barneygale commented:
Jason, this might be a good opportunity for me to say "thank you" for `zipp`! It's a gem of a package and it was the [original inspiration](https://discuss.python.org/t/make-pathlib-extensible/3428/1) for my work on pathlib ~5 years ago, which led to me becoming a CPython triager and then a developer. Thank you for the splendid inspiration :-)

I hope I can convince you that this PR is a good idea. To summarize the improvements:

1. It provides a `glob()` implementation that's 100% compatible with stdlib `pathlib`/`glob`.
2. It solves all issues around trailing slashes for directories. For example, `ZipPath(..., at=at).is_dir()` now returns the correct value whether or not `at` has a trailing slash!
3. It adds a few attributes/methods that are useful to users, with more available but disabled
   - We could enable `walk()`, which is pretty handy!
   - `copy()` and `copy_into()` would allow zip files to be extracted with `zipp.Path(...).copy(pathlib.Path(...))` (3.14+ only)
4. It's a stepping stone towards full write support
   - We could also subclass `WritablePath` and add `mkdir()` and `symlink_to()` methods
   - This would allow zip files to be populated with `pathlib.Path(...).copy(zipp.Path(...))` (3.14+ only)
   - Copies between zip files can be done with `zipp.Path(...).copy(zipp.Path(...))`
5. It removes ~250 lines of code!

I've tried to keep the changes minimal. I'd love to hear what you think, when you have time. Thanks again!

jaraco commented on zipp/__init__.py:
```diff
@@ -7,204 +7,83 @@
+    def __init__(self, items=tuple(), exists=True):
+        self._exists = exists
+        self.zip_info = None
+        self.children = {}
+        for zip_info in items:
+            self.append(zip_info)
```
Do I understand correctly that this new paradigm will eagerly build the tree of PathInfo objects for the whole zipfile? Could that be problemmatic? What about for use-cases where the same zipfile is used on different Path objects? Could it prove costly that ZipInfo objects are constructed eagerly?

barneygale replied:
> Do I understand correctly that this new paradigm will eagerly build the tree of PathInfo objects for the whole zipfile?

That's correct

> Could that be problemmatic? What about for use-cases where the same zipfile is used on different Path objects?

That works fine I think. In `zip.Path.__init__()`, we only replace the zipfile's `filelist` if it's not already an instance of `PathInfo`

> Could it prove costly that ZipInfo objects are constructed eagerly?

(We don't eagerly construct ZipInfo objects I don't think! Presuming you meant PathInfo objects:) Yes there's a cost - for every part of every member path, we perform a dictionary lookup and usually an insertion.

A call like `zipp.Path('foo/bar/').is_dir()` will generally be slower because it performs two lookups (one for each level) rather than one. Though we lose a little in performance we gain a little in correctness, because `Path('foo//bar').is_dir()` will now work correctly.

A call like `zipp.Path('foo/bar/').iterdir()` should be a little faster because the directory children are already recorded on the `PathInfo` instance, so we don't need to iterate over the entire list of members as in the old implementaiton.

jaraco replied:
SGTM. Thanks for the clarification. I do think this is a strong improvement.

jaraco commented on zipp/__init__.py:
```diff
@@ -7,204 +7,83 @@
+        return length
+
+    def append(self, zip_info):
+        self.resolve(zip_info.filename, True).zip_info = zip_info
```
```suggestion
        self.resolve(zip_info.filename, create=True).zip_info = zip_info
```

jaraco reviewed (commented):
This is great. I love what you're thinking here, and I love the elegance that comes out of it (especially in the Path class). Let's work through the details and some of the questions, but I'm confident we can get something very similar to this merged.

jaraco commented:
I've pushed a few changes to address some of the nitpicky linter and other obvious fixes. I've saved those changes in a separate branch, so it's no problem if you wish to re-write anything and force push.

I've also noticed there are some missed lines in the coverage:

```
Name                       Stmts   Miss  Cover   Missing
--------------------------------------------------------
conftest.py                    4      0   100%
docs/conf.py                  15      0   100%
tests/__init__.py              0      0   100%
tests/_support.py              5      0   100%
tests/_test_params.py         21      0   100%
tests/compat/__init__.py       0      0   100%
tests/compat/py38.py           2      0   100%
tests/compat/py39.py           4      0   100%
tests/compat/py310.py          4      0   100%
tests/test_complexity.py      52      1    98%   30
tests/test_path.py           435      3    99%   193-195
tests/write-alpharep.py        2      0   100%
zipp/__init__.py             154      4    97%   81, 245, 309, 314
zipp/compat/__init__.py        0      0   100%
zipp/compat/overlay.py        11      0   100%
zipp/compat/py310.py           3      0   100%
--------------------------------------------------------
TOTAL                        712      8    99%
```

It would be nice to get back to 100%.

jaraco commented:
One failure I see when I build locally is test_encoding_warnings:

```
____________________________________________________________ TestPath.test_encoding_warnings ____________________________________________________________

self = <tests.test_path.TestPath testMethod=test_encoding_warnings>, alpharep = <zipfile.ZipFile file=<_io.BytesIO object at 0x1057ceb60> mode='w'>

    @unittest.skipIf(
        not getattr(sys.flags, 'warn_default_encoding', 0),
        "Requires warn_default_encoding",
    )
    @pass_alpharep
    def test_encoding_warnings(self, alpharep):
        """EncodingWarning must blame the read_text and open calls."""
        assert sys.flags.warn_default_encoding
        root = zipfile.Path(alpharep)
        with self.assertWarns(EncodingWarning) as wc:  # noqa: F821 (astral-sh/ruff#13296)
            root.joinpath("a.txt").read_text()
>       assert __file__ == wc.filename
E       AssertionError: assert '/Users/jarac.../test_path.py' == '/Users/jarac...ib_abc/_os.py'
E         
E         - /Users/jaraco/code/jaraco/zipp/.tox/py/lib/python3.13/site-packages/pathlib_abc/_os.py
E         + /Users/jaraco/code/jaraco/zipp/tests/test_path.py

tests/test_path.py:192: AssertionError
```

It's important that the stack level is set correctly so it warns at the the right location.

jaraco commented on zipp/__init__.py:
```diff
@@ -434,22 +320,27 @@ def relative_to(self, other, *extra):
 
     def __str__(self):
-        return posixpath.join(self.root.filename, self.at)
+        return self.at
```
This change is causing one of the test failures:

```
__________________________________________________________________ [doctest] zipp.Path __________________________________________________________________
141 
142 >>> c.exists()
143 True
144 >>> (b / 'missing.txt').exists()
145 False
146 
147 Coercion to string:
148 
149 >>> import os
150 >>> str(c).replace(os.sep, posixpath.sep)
Expected:
    'mem/abcde.zip/b/c.txt'
Got:
    'b/c.txt'
```

barneygale replied:
> Why was this changed?

The `JoinablePath` base class calls `str(self)` to obtain a string representation of the virtual path, which it feeds into various `posixpath` functions in order to implement `.parent`, `with_suffix()`, etc. For this to work correctly, `zipp.Path.__str__()` needs to return _just_ the member path without the `ZipFile.filename` prefix.

(It's possible to make `JoinablePath` handle the zip file filename prefix by defining a custom `PathParser` rather than using `posixpath`, but it's a PITA and not worth the trouble IMO.)

Putting aside the pathlib-abc stuff, IMHO it's still preferable for `__str__()` to return just the member path. I logged #134 about that.

jaraco replied:
I disagree, maybe strongly. The `Path` object is about much more than just a path. It's a about a path within a zipfile. The user cares about that context and the holder of that object may no longer have a handle to the original zipfile, so the `str` should represent it. I'm certain I've relied on this functionality and I'd be surprised if no one else does.

If the motivation for making this change is for convenience of the internal implementation, that feels like the wrong tradeoff. If on the other hand, we want the user's experience to be oblivious to the attached zipfile, that's different.

I do think I see the conflict here. I do  think I'm leaning toward providing a custom `PathParser`, especially if that's the only option.

You know what I'd like better is to revisit this assumption.



> The `JoinablePath` base class calls `str(self)` to obtain a string representation of the virtual path

What if instead `JoinablePath` were to call `self.__path__` (or some other special hook) to obtain the joinable path (falling back of course to `self.__str__`). That would give classes like `zipp.Path` the ability to provide the short path for joining, but avoid the need for its `__str__` to have a special representation? I think I'd even go a step further and make `JoinablePath` _require_ that `__path__` (or similar) to be implemented and not have _any_ reliance on `__str__`, which is implicated in other important user-facing behaviors.

jaraco replied:
It sounds like we're on track to restoring the expectation for `__str__`. I'm looking forward to that.

barneygale commented on zipp/__init__.py:
```diff
@@ -334,28 +218,35 @@ def __eq__(self, other):
     def __hash__(self):
         return hash((self.root, self.at))
 
+    def __reduce__(self):
+        return (self.__class__, (self.root.filename, self.at))
```
Right! Thanks for the explanation - I didn't quite understand what `InitializationState` was doing. I'll restore previous behaviour.

BTW what's the use case for pickleable `zipp.Path` objects? It took me a bit by surprise because obviously `zipfile.ZipFile` isn't pickleable

barneygale replied:
I've "fixed" this by recording the initial argument as `Path._initial_arg` and propagating that to derived paths. I guess I don't really see the point of turning it into a whole class with its own backing module - isn't that something of an over-generalisation of something used only once?

jaraco replied:
> BTW what's the use case for pickleable `zipp.Path` objects? It took me a bit by surprise because obviously `zipfile.ZipFile` isn't pickleable

#81 describes the motivation.

> I've "fixed" this by recording the initial argument as `Path._initial_arg` and propagating that to derived paths. I guess I don't really see the point of turning it into a whole class with its own backing module - isn't that something of an over-generalisation of something used only once? Perhaps I'm missing something

My motivation for creating the standalone class was to separate concerns. By making it an independent class, it allows that functionality to be selectively added or removed, and it makes it easier to see what bits affect Pickleability (instead of reading through the whole code of Path to try to infer which bits are relevant). I try to avoid entangling concerns wherever possible. It also makes it easier to detect when that behavior is altered (helps avoid unintentional regressions). It does also have the benefit of potentially being generalizable and reusable, which could prove useful should another project ask for something similar.

barneygale replied:
Gotcha, thanks for the extra detail. I've restored the `InitializedState` implementation.

But the proposed `zipp.Path` implementation doesn't use `CompleteDirs` nor `FastLookup`, and I don't think it can inherit `InitializedState` directly. So I think I still need a `Path.__reduce__` method, unless I'm missing something

jaraco replied:
There's little value in retaining the `InitializedState` if it's not somehow inherited by the `Path` class. I wonder why `Path` couldn't inherit from `InitializedState`. If that class can't be used to add pickleability to `zipfile.ZipFile` (or whatever property of `Path` that isn't pickleable), then maybe it's not usable. It gets complicated because `Path` encapsulates `ZipFile`, so pickleability concerns are entangled. I'll take another look.

barneygale replied:
> I wonder why `Path` couldn't inherit from `InitializedState`.

The problem comes when you want to create derived paths, e.g. from `Path.parent` or `Path.iterdir()`. Those methods call `with_segments()` (previously `_next()`). That method calls something like:

```python
return self.__class__(self.root, at)
```

Where `self.root` is a `ZipFile` object, _not_ a filename string, because you don't want to re-open the zip file with every derived path. But if `zipp.Path` were to inherit `InitializedState`, then only the `ZipFile` object (not the `str`) would be captured, and so any derived paths wouldn't be pickleable.

jaraco replied:
> > There's little value in retaining the `InitializedState` if it's not somehow inherited by the `Path` class.
> 
> The only fly in the ointment is that `zipfile` exposes `CompleteDirs`, which inherits `InitializedState`, so arguably we need to go through a deprecation period first

Then let's do that. The only reason I ever exported that was for "internal" use (maybe in other libraries like importlib_metadata or importlib_resources).

I still need to look at this again.

barneygale commented:
I've opened a CPython PR to fix `test_encoding_warnings`: https://github.com/python/cpython/pull/133051

barneygale commented on zipp/__init__.py:
```diff
@@ -7,204 +7,83 @@
-    Multiple separators are treated like a single.
-
-    >>> list(_ancestry('//b//d///f//'))
-    ['//b//d///f', '//b//d', '//b']
```
I've restored this function

barneygale commented on tests/test_path.py:
```diff
@@ -72,7 +72,7 @@ def build_alpharep_fixture():
 alpharep_generators = [
     Invoked.wrap(build_alpharep_fixture),
-    Invoked.wrap(compose(zipfile._path.CompleteDirs.inject, build_alpharep_fixture)),
+    #Invoked.wrap(compose(zipfile._path.CompleteDirs.inject, build_alpharep_fixture)),
```
I've restored this line and `CompleteDirs` implementation, though it's not used by `Path`.

`CompleteDirs` isn't listed in `__all__`, fwiw

barneygale commented on zipp/__init__.py:
```diff
@@ -7,204 +7,83 @@
 
-from .compat.py310 import text_encoding
-from .glob import Translator
+import pathlib_abc
```
The ABCs are currently private in CPython, so it's a bit iffy for `zipp` to use the stdlib at the moment.

That could all change if I get a PEP accepted that makes them public, but I think that's a ways off yet. For now could `zipp` always use the `pathlib-abc` versions?

I'll adopt semver in `pathlib-abc` and align major versions with cpython minor versions shortly.

jaraco replied:
Good question - this project has an (unstated, implied) strict policy that it's portable to the stdlib. We'll want to either (a) stabilize or at least merge the stdlib implementation or (b) vendor the pathlib-abc functionality when porting to CPython. In either case, I'd like for this line to be a single line that's applicable to both implementations (even if that line mutates across the implementations). It's not urgent that we solve this portability concern now, but it is important that before the release of Python 3.14 b1 that there's a clear strategy for incorporating this functionality into CPython (or delay to Python 3.15+). Let's not fuss about it now; I can manage the portability concerns.

jaraco commented:
I took a quick refresh of this review. It looks like it's progressing nicely. I may have time this week to work on this more. It's top of my list.

barneygale commented:
> I took a quick refresh of this review. It looks like it's progressing nicely. I may have time this week to work on this more. It's top of my list.

No rush at all mate, this is a big change and I understand completely if you can only review occasionally

barneygale commented:
Here's a CPython PR that replaces `__str__()` with `__vfspath__()` in the pathlib ABCs: https://github.com/python/cpython/pull/133437

barneygale commented on pyproject.toml:
```diff
@@ -18,6 +18,7 @@ classifiers = [
 ]
 requires-python = ">=3.9"
 dependencies = [
+	"pathlib-abc == 0.4.1"
```
@jaraco As you say, in CPython it should be OK because you can pull in the relevant stuff from `pathlib.types` and `pathlib._os`.

But I think you might have a point about the PyPI package. `zipp` is core packaging infrastructure, and if it depends on `pathlib-abc` then that becomes core infra too. But the testing/publishing setup isn't as sophisticated as `zipp`'s, and you'd be dependent on me to release new versions as you need.

Just a thought: would you be interested in co-maintaining `pathlib-abc`? It would be terrific if you can bring it up to the same standard as your other packages, but I understand completely if you'd rather not, or don't have the time.

jaraco replied:
Good call. This sounds right to me. I'd be happy to co-maintain pathlib-abc. I have a whole system ([two][jaraco/skeleton] [actually](/coherent-oss/system)). If you might be interested in the Coherent System, let me know and I'll prepare a PR showing how much boilerplate you don't need. In the meantime, let's add the dependency here. Let's just make sure this project has a compatibility shim that gets replaced by stdlib `pathlib._os` when merged into the `cpython` branch (I can do the merge - let's just keep the diff as small as possible on the shared code).

jaraco commented on zipp/__init__.py:
```diff
@@ -195,17 +193,73 @@ def _name_set_prop(self):
         return super()._name_set()
 
 
-def _extract_text_encoding(encoding=None, *args, **kwargs):
-    # compute stack level so that the caller of the caller sees any warning.
-    is_pypy = sys.implementation.name == 'pypy'
-    # PyPy no longer special cased after 7.3.19 (or maybe 7.3.18)
-    # See jaraco/zipp#143
-    is_old_pypi = is_pypy and sys.pypy_version_info < (7, 3, 19)
-    stack_level = 3 + is_old_pypi
-    return text_encoding(encoding, stack_level), args, kwargs
```
When I test the change on my local machine, running PyPy 7.3.17, I get a test failure because of the removal of this logic.

```
____________________________________________________________ TestPath.test_encoding_warnings ____________________________________________________________

self = <tests.test_path.TestPath testMethod=test_encoding_warnings>
alpharep = <zipfile.ZipFile file=<_io.BytesIO object at 0x00000001432c1a48> mode='w'>

    @unittest.skipIf(
        not getattr(sys.flags, 'warn_default_encoding', 0),
        "Requires warn_default_encoding",
    )
    @pass_alpharep
    def test_encoding_warnings(self, alpharep):
        """EncodingWarning must blame the read_text and open calls."""
        assert sys.flags.warn_default_encoding
        root = zipfile.Path(alpharep)
        with self.assertWarns(EncodingWarning) as wc:  # noqa: F821 (astral-sh/ruff#13296)
            root.joinpath("a.txt").read_text()
>       assert __file__ == wc.filename
E       AssertionError: assert '/Users/jarac.../test_path.py' == '/Users/jarac...c/__init__.py'
E         
E         - /Users/jaraco/code/jaraco/zipp/.tox/pypy/lib/pypy3.10/site-packages/pathlib_abc/__init__.py
E         + /Users/jaraco/code/jaraco/zipp/tests/test_path.py

tests/test_path.py:193: AssertionError
```

See #143 where I observed that the idiosyncratic PyPy behavior changed recently.

Maybe it's not important and this special handling can be dropped, but we should be explicit about the behavior (i.e. skip failing tests on older PyPy versions).

barneygale replied:
I reckon this is the pypy change: https://github.com/pypy/pypy/commit/6ca2f486ed1587a4d0fa693ce011e57c78e7d6c8

IMHO the bug is too minor and too obscure to be worth working around in zipp. I'll adjust the tests as you suggest

barneygale committed:
Add dependency on `pathlib-abc`

Make `zipp.Path` subclass `pathlib_abc.ReadablePath`. This allows us to
remove implementations of `read_text()`, `read_bytes()`, `glob()`,
`joinpath()` and `__truediv__()`, and to simplify implementations of a
couple of few more methods.

Maintain a tree of `PathInfo` objects representing the hierarchy of zip
file members. We traverse the tree whenever we need to resolve a path to
a `ZipInfo` object. This effectively hides the `.zip`-specific quirk that
directories are recorded with a trailing slash in their filenames.

Adjust `__str__()` so that it doesn't raise for unnamed zip files. Instead
it returns a string beginning with `:fileobj:`. The prefix is made
available as a new `anchor` attribute.

Add the following methods/attributes, which are required by `ReadablePath`:

- `info`
- `parser`
- `__open_rb__()`
- `readlink()`
- `with_segments()`

Disable the following `ReadablePath` methods/attributes that we don't (yet) test in zipp:

- `anchor`
- `parts`
- `parents`
- `__rtruediv__()`
- `with_name()`
- `with_stem()`
- `with_suffix()`
- `full_match()`
- `walk()`
- `copy()`
- `copy_into()`

barneygale committed:
Skip test in old PyPy versions

barneygale committed:
Lint

barneygale commented:
Hey @jaraco, I think this is reviewable now.

I've published a new version of pathlib-abc that adds a `JoinablePath.__vfspath__()` method, which replaces `__str__()` as an abstract method. Consequently this patch no longer makes any changes to `zipp.Path.__str__()`.

I haven't yet made any zipp changelog/docs changes. Let me know when you think they're due.

Coverage is reduced because `InitializedState`, `CompleteDirs`, `FastLookup`, `save_method_args`, `Translator` and `separator` are no longer used by `Path`. I'm not sure whether to write new tests for them or remove them from `zipp` - let me know what you thinks best?

jaraco commented:
> It would be good to land #149 first, as it would simplify this patch.

That's landed now. Let me know if you can resolve the merge conflicts. Alternately, I can look at some point (sooner than months ;)
